### PR TITLE
sort tiles

### LINKED ISSUE #622: Make sorting available for Game tiles
Make it possible to sort tiles

### PR DESCRIPTION
**Pull request by axel1200**
Fixes #622

### CODE REVIEW

**Comment by axel1200** [comment]
In the Method toLowerCase()
Using US Local as Games Titles are supposingly in English

File: src/main/java/com/faforever/client/game/CustomGamesController.java [hunk 1]
```diff
@@ -60,19 +67,37 @@
...
     this.preferencesService = preferencesService;
     this.eventBus = eventBus;
+    this.i18n = i18n;
   }
 
   public void initialize() {
+    choseSortingTypeChoiceBox.setVisible(false);
+    Map<SortingOrder, String> sortingOrderToString = new HashMap<>();
+    Arrays.stream(SortingOrder.values())
+        .forEach(sortingOrder -> sortingOrderToString.put(sortingOrder, i18n.get(sortingOrder.getDisplayNameKey())));
```
Comment (micheljung) [inline src/main/java/com/faforever/client/game/CustomGamesController.java]: Map<SortingOrder, String> sortingOrderToString = Arrays.stream(SortingOrder.values())
        .collect(Collectors.toMap(Function.identity(), sortingOrder -> i18n.get(sortingOrder.getDisplayNameKey())));

File: src/main/java/com/faforever/client/game/CustomGamesController.java [hunk 2]
```diff
@@ -60,19 +67,37 @@
...
+    choseSortingTypeChoiceBox.setConverter(new StringConverter<SortingOrder>() {
+      @Override
+      public String toString(SortingOrder sortingOrder) {
+        return sortingOrderToString.get(sortingOrder);
+      }
+
+      @Override
+      public SortingOrder fromString(String string) {
+        return sortingOrderToString.entrySet().stream().filter(sortingOrderStringEntry -> sortingOrderStringEntry.getValue().equals(string)).findFirst().get().getKey();
```
Comment (micheljung) [inline src/main/java/com/faforever/client/game/CustomGamesController.java]: As we don't convert from String to SortingOrder, you can leave this empty

File: src/main/java/com/faforever/client/game/CustomGamesController.java [hunk 3]
```diff
@@ -60,19 +67,37 @@
   public Pane gamesRoot;
   public ScrollPane gameDetailPane;
   public GameDetailController gameDetailController;
+  public ChoiceBox<SortingOrder> choseSortingTypeChoiceBox;
```
Comment (micheljung) [inline src/main/java/com/faforever/client/game/CustomGamesController.java]: [choose](https://www.dict.cc/?s=choose)

File: src/main/java/com/faforever/client/game/GamesTilesContainerController.java [hunk 4]
```diff
@@ -27,14 +35,19 @@
 
   private final UiService uiService;
   private final ListChangeListener<Game> gameListChangeListener;
+  private final PreferencesService preferencesService;
   public FlowPane tiledFlowPane;
   public ScrollPane tiledScrollPane;
-  private Map<Integer, Node> uidToGameCard;
+  @VisibleForTesting
+  protected Map<Integer, Node> uidToGameCard;
```
Comment (micheljung) [inline src/main/java/com/faforever/client/game/GamesTilesContainerController.java]: Should be package private, not protected.

File: src/main/java/com/faforever/client/game/GamesTilesContainerController.java [hunk 5]
```diff
@@ -43,24 +56,48 @@ public GamesTilesContainerController(UiService uiService) {
...
 
+  private void sortNodes() {
+    ObservableList<Node> sortedChildren = tiledFlowPane.getChildren().sorted(appliedComparator);
+    tiledFlowPane.getChildren().setAll(sortedChildren);
+  }
+
   public void initialize() {
     JavaFxUtil.fixScrollSpeed(tiledScrollPane);
+
+    sortingListener = (observable, oldValue, newValue) -> {
```
Comment (micheljung) [inline src/main/java/com/faforever/client/game/GamesTilesContainerController.java]: As this sortingListener never changes, it should be `final` and initialized in the constructor.

File: src/main/java/com/faforever/client/game/GamesTilesContainerController.java [hunk 6]
```diff
@@ -43,24 +56,48 @@ public GamesTilesContainerController(UiService uiService) {
...
+      sortNodes();
+    };
   }
 
   ReadOnlyObjectProperty<Game> selectedGameProperty() {
     return this.selectedGame;
   }
 
-  void createTiledFlowPane(ObservableList<Game> games) {
+  void createTiledFlowPane(ObservableList<Game> games, ChoiceBox<SortingOrder> choseSortingTypeChoiceBox) {
```
Comment (micheljung) [inline src/main/java/com/faforever/client/game/GamesTilesContainerController.java]: `@VisibleForTesting`

File: src/main/java/com/faforever/client/game/GamesTilesContainerController.java (hunk 4 above)
Comment (axel1200) [reply src/main/java/com/faforever/client/game/GamesTilesContainerController.java]: But then it is no longer VisibleForTesting!!!!

Comment (micheljung) [reply src/main/java/com/faforever/client/game/GamesTilesContainerController.java]: Yes it is. I said `package private` not `private`. Means, no modifier, means, the field will be visible from the same package. And unit tests are in the same package :)

**Pull request merged**